### ISSUE TITLE
Ufunc calls on scalars are very slow

### ISSUE THREAD

**Issue by mhvk** [numpy/numpy#11232]
It is well known that ufunc calls on scalars are rather slow, but it is probably good to have a summary of why it is so slow, for which it is useful to go along the `ufunc_generic_call` path. I got only partway, but one possible solution might be for the scalars to already get overridden in in `CheckOverride`, i.e., to treat them as if they had their own `__array_ufunc__` (with priority even below that of ndarray; an actual `__array_ufunc__` calling `math` is slightly slower than our present state).

1.  `PyUFunc_CheckOverride`: for non-arrays (thus including scalars), this checks whether the scalar has `__array_ufunc__`. Easy to avoid if our whole API is available - needs #10915.
2. `PyUFunc_GenericFunction`: to be done (will edit).
3. `make_full_arg_tuple`: for no output arguments, needlessly creates and destroys a tuple (fix in #11231).
4. `_find_array_wrap` -> `_find_array_method`: skips arrays and scalars, so should be reasonably fast (though a subclass check for `Generic` is done before type checks on python objects in `PyArray_IsAnyScalar` (in `ndarrayobject.h`).

**Comment by eric-wieser** [comment]
Does this have a benchmark? If not, would be good to add one before digging any further

**Comment by seberg** [comment]
IIRC one of the slowest things right now may be the type resolution loop of the ufunc, because I remember discussions that implementing a hash table for it would be good, and I somewhat don't think that has been done. A few years back there were some optimizations done in the scalar paths in a GSoC I think, so a lot of the lowest hanging fruits are likely gone (though new ones might have come up I guess).

**Comment by mhvk** [comment]
@seberg - I'm mostly looking to see if my idea of intercepting already at the `CheckOverride` is worth it - if, say, a single-input ufunc gets just a double, can we speed things up. In some sense, the approach would be to thinkt that, as is, everything without `__array_ufunc__` is considered an array, but one might as well think of scalars as having their own implied `__array_ufunc__`, which we known can deal with scalars only.

**Comment by seberg** [comment]
Yeah, it sounds like a good idea, a few simple fast paths like that for scalars could go a far way. If we can get such a thing right in a not too complex way, maybe it can even help with reducing the code complexity/duplication that is currently really annoying for scalars? (I believe in scalars, but that extra code I do not believe in)
There are also some really slow things probably such as buffering/casting necessary for things such as `np.add(1., 3)` which likely is hilariously slow....

**Comment by mhvk** [comment]
@eric-wieser - good point about the benchmarks. From a quick look, it looks like there are benchmarks for array scalars but not for true scales (neither numpy nor python ones).

**Comment by mhvk** [comment]
@seberg - there is room for improvement... I now added a few timings on top.

**Comment by juliantaylor** [comment]
here is one of the type lookup optimizations:
https://github.com/numpy/numpy/pull/4904

**Comment by mhvk** [comment]
Still would be good to see if one cannot optimize ufuncs called on scalars a bit. Timings on numpy-dev (nearly 2.0)
```
a = 1.
a64 = np.float64(1.)
as64 = np.array(1., dtype=np.float64)
%timeit math.sin(a)  # and a64, as64
# 53, 58, 70  ns for a, a64, as64
%timeit np.sin(a)
# 1010, 920, 560 ns for a, a64, as64
```
